### PR TITLE
Manual loss & chained setting

### PR DESCRIPTION
**Pull request by psinger**
This PR addresses to points:

First, it introduces a new setting `limit_chained_samples` which limits the chained samples to the longest one. Please feel free to suggest a better name for that.

Second, it reworks loss calculation by adding functionality to calculate the loss in our code base instead of relying on the one in `transformers` model class. This will also allow us long-term more flexibility, in terms of adding new loss functions, weighted loss, etc.

For now, there are two BCE loss variations: 
- `TokenCrossEntropy` calculates the loss as before on token level
- `SampleCrossEntropy` calculates the loss separate per sample in a batch, and then averages per sample

Experiments seem to indiciate that in generel the old way works better.

### CODE REVIEW

File: llm_studio/src/models/text_causal_language_modeling_model.py [hunk 1]
```diff
@@ -193,11 +195,12 @@ def forward(
             output = self.backbone(
                 input_ids=batch["input_ids"],
                 attention_mask=batch["attention_mask"],
-                labels=batch["labels"],
+                labels=None,
```
Comment (pascal-pfeiffer) [inline llm_studio/src/models/text_causal_language_modeling_model.py:198]: or remove line entirely to avoid confusion? None is the default, right?

**Review by pascal-pfeiffer — APPROVED** [review]
Thank you, lgtm!

Two comments in the code. Your choice to address them or merge as is.
